### DISCUSSION TITLE
[typegen] transition between children of compound states

### DISCUSSION THREAD

**Discussion by EightArmCode** [statelyai/xstate#3404] [Q&A]
We would like to refer to the child state of a compound state from the child state of a (previous) compound state. Example machine: 

```
const machine = {
  id: 'myMachine',
  initial: 'A',
  states: {
    A: {
      id: 'A',
      initial: 'A1',
      states: {
        A1: {
          on: {
            GO_TO_B1: '#myMachine.B',
          },
        },
        A2: {
          on: {
            GO_TO_B2: '#B.B2'
          }
        },
      },
    },
    B: {
      id: 'B',
      initial: 'B1',
      states: {
        B1: {
          on: {
            GO_TO_A1: '#myMachine.A'
          }
        },
        B2: {
          on: {
            GO_TO_A2: '#A.A2'
          }
        },
      },
    },
}
```

We added IDs to the compound state nodes, and it generates types when we reference the child state with syntax `#myMachine.A`, however we cannot reference the state we want to target with `#A` or `#B.B2`. We also cannot access `#myMachine.B.B2`. Ultimately the transition we want would be to go from one substate to the other. If current state was `#myMachine.B.B2` and we wanted to transition to `#myMachine.A.A1`, then typegen complains with:

```
Error: Invalid transition definition for state node...
Child state node '#A' does not exist on machine 'myMachine'

```

Since it is a previous state, we looked into the docs for history nodes, but felt discouraged by the note at the bottom:

>History states can be directly accessed from State instances on state.history, but this is seldom necessary.

We added a [codesandbox](https://codesandbox.io/s/practical-bird-eiyq4g?file=/src/App.tsx). Hopefully the transitions are invoked correctly to demonstrate the problem

_Disclaimer: we translated proprietary code to pseudocode for this discussion._

**Comment by Andarist** [comment]
I think that the example transitions are incorrect - you most likely have meant those:
```diff
-machine.transition("#myMachine.A.A2", "GO_TO_A2");
-machine.transition("#B.B2", "GO_TO_B2");
-machine.transition("#A.A2", "GO_TO_A2");
+machine.transition("#myMachine.A.A2", "GO_TO_B2");
+machine.transition("#B.B2", "GO_TO_A2");
+machine.transition("#A.A2", "GO_TO_B2");
```

It doesn't quite matter for the issue though. It seems that we convert the first argument given to `machine` differently than when executing transitions - you are right that using `machine.transition` like this throws. However, when using live services it works just OK:
https://codesandbox.io/s/blue-sky-5j7bsl?file=/src/App.tsx

This is definitely something that we should fix and I will get to this soon. Out of curiosity - are you using `machine.transition` API in your real project? Why are you using it over the live interpreter?

**Reply by EightArmCode** [reply]
We're actually using the `@xstate/react useInterpreter`. The problem occurs on the lines where we specify the transitions. Like so: 

```typescript
states: {
  B1: {
    on: {
      GO_TO_A2: #myMachine.A.A2. // typegen complains that it can't find state A on myMachine
    }
  }
}
```

I used `.transition` in the sandbox because it was the most simple way to show the problem, and seems to be the convention in other examples I found.

**Reply by EightArmCode** [reply]
We tried targeting the state with an ID, so changing the offending line to `GO_TO_A2: #A.A2` and got the same error.

FWIW the history node seems to suit our use case, but I am curious if there might be lost context changes, etc?

**Reply by Andarist** [reply]
> The problem occurs on the lines where we specify the transitions. Like so:

Could you take a look at the codesandbox that I've provided? The logs at those lines seem to be correct and there is no error there:
https://codesandbox.io/s/blue-sky-5j7bsl?file=/src/App.tsx:714-876

There must be some difference between this and your code - or I'm misunderstanding what the issue is about.

> typegen complains that it can't find state A on myMachine

I'm not sure how typegen relates to this problem because it seems that it's a runtime problem.

> FWIW the history node seems to suit our use case, but I am curious if there might be lost context changes, etc?

Hard to tell what you are trying to achieve here based on the given example. History states are helpful to get back to some state configuration from the time that a particular parent state was exited for the very last time. All entry actions etc should still be executed within "restored" state there but your preexisting context etc will be kept (it might get changed by those entry actions though, but those would be "additive" changes)

**Comment by EightArmCode** [comment]
It appears you did improve the sandbox, but it is not a runtiume/compilation error. Typegen specifically errors on the line where state B is defined, from the initial state of that compound child state. There is a transition defined that I mentioned above:

```typescript
states: {
  B1: {
    on: {
      GO_TO_A2: #myMachine.A.A2 // typegen complains that it can't find state A on myMachine
    }
  }
}
```

It does not find state A, whether referenced by `#myMachine.A.A2` or if we give state A an ID and try to reference it with `#A.A2`.

I should probably also mention that state A is defined in another file.

**Reply by Andarist** [reply]
If the problem involves typegen then please prepare a repro case that will use the typegen. It can be a simple repo that I could clone and investigate
